The update of UCDavisService inculdes:

DhYS opened the pull request:
1.Add null check void method in writeToCSV
2.check the map with containsKeys(dailyKeys and hourlyKeys):
	If key exists: check the values, fill the values as "0.0" if values don't exist
	If key doesn't exist: add the key, read the dateKeys and add mapping dates with "-1"
3.finish write to csv

DhYS committed:
The update of UCDavisService inculdes:

1.Add null check void method in writeToCSV
2.check the map with containsKeys(dailyKeys and hourlyKeys):
	If key exists: check the values, fill the values as "0.0" if values don't exist
	If key doesn't exist: add the key, read the dateKeys and add mapping dates with "-1"
3.finish write to csv

chrisyly commented on ucdavis_service/src/main/java/com/ucdavis/application/UCDavisService.java:
```diff
@@ -140,6 +140,47 @@ public void writeToMap(List<JSONObject> records, String target) {
+		// TODO hourly keys
+		String[] hourlyKeys = {};
+		
+		SortedSet<String> sortedDateKeys = new TreeSet<String>(dataItems.get("Date").keySet()); // dates cant be null
```
How you check the null date?

DhYS replied:
For CIMIS service, the date won't be null, that's why I put the key "Date" search here. So should I assume the case exists that dates could be null?

chrisyly commented on ucdavis_service/src/main/java/com/ucdavis/application/UCDavisService.java:
```diff
@@ -140,6 +140,47 @@ public void writeToMap(List<JSONObject> records, String target) {
+		
+		// check the null keys and values
+		
+		String[] keys = isDaily? dailyKeys : hourlyKeys;
```
This is good

chrisyly commented on ucdavis_service/src/main/java/com/ucdavis/application/WebController/ServiceController.java:
```diff
@@ -66,6 +66,7 @@ public String submit(@Valid @ModelAttribute("request") final Request request,fin
 		List<JSONObject> list = parser.getListFromJson(parser.extractDataItems((parser.jsonStringToMap(jsonObject)).get("Records").toString()));
 		// write map
 		davisService.writeToMap(list, queryMapping.get("targets"));
+		davisService.writeToCSV(queryMapping.get("targets"));
```
Please check if write to CSV from map is also working

DhYS replied:
WriteToMap is based on the map null-checked, it is working now.

DhYS replied:
WriteToCSV is based on the HashMap after null check. It has been tested.

chrisyly commented on ucdavis_service/src/main/java/com/ucdavis/application/UCDavisService.java:
```diff
@@ -140,6 +140,47 @@ public void writeToMap(List<JSONObject> records, String target) {
+		checkNullKeysAndValue(dataItems, true);
+	}
+	
+	private void checkNullKeysAndValue(Map<String, Map<String,String>> dataItems, boolean isDaily) {
```
suggest to have a general map null check, let me know your opinion on this. The general null check is basically taking any input map, it should work exactly like how we read the query and put it into the map

DhYS replied:
So you suggest that the input should set as like Map<String, Obejct>,... ?

chrisyly reviewed (commented):
Please review the comments and update. if it is possible make a makefile or build script (like run.sh)

DhYS committed:
User story of WriteToMap update on Oct.7th
1. Function used: writeToMap(main), fillMissingDateKeys, checkNullKeysAndValue, addOneDay
2. Input: List<JSONObject> records, Request request(contains start date, end date and targets)
3. Output: write to local HashMap(Map<String, Map<String, Map<String, String>> data)
4. In writeToMap: there are 3 cases for List<JSONObject> records to consider about
	i.  records == null, the input JSONObject list is null
	ii. the dates in records is not complete:
		example: data includes Oct.1st, Oct.2nd, Oct.4th, Oct.5th(Oct.3rd is null)
	iii.the sequence of dates are continuous but some value are missing
 5. The scope of the writeToMap:
    - create a Map<String, Map<String, String>> dataItems which will be the value the key of corresponding target,
	 	For example, In data of Step 3: “2”(target is 2) —> dataItems
    - if records is null, call the function fillMissingDateKeys to fill in the keys and values about “Date” in the dataItems according the start_date and end_date stored in Java variable Request
		For example, 2 —>{ “Date” —>{“Oct.1st=Oct.1st”, “Oct.2nd=Oct.2nd”, “Orc.3rd=Oct.3rd”,…}}
    - if records is not null, but dates are not completed, write the existing keys and values in dateItems
		For example, dateItems will be like: 2 —>{ “Date” —>{“Oct.1st=Oct.1st”, “Oct.2nd=Oct.2nd”, “Orc.4th=Oct.4th”,…}, “Temperature”—>{“Oct.1st=23”, “Oct.2nd=14”, “Oct.4th=27”,…} }
    - then call call the function fillMissingDateKeys, dataItems will be changed to:
		2 —>{ “Date” —>{“Oct.1st=Oct.1st”, “Oct.2nd=Oct.2nd”, ”Oct.3rd=Oct.3rd” “Oct.4th=Oct.4th”,…}, “Temperature”—>{“Oct.1st=23”, “Oct.2nd=14”, “Oct.4th=null”,…} }, which Orc.3rd is added
    - finally call the checkNullKeysAndValue to fill up the missing values, if column not exists, fill as -1, if value is null, fill as 0.0
		example: 2 —>{ “Date” —>{“Oct.1st=Oct.1st”, “Oct.2nd=Oct.2nd”, ”Oct.3rd=Oct.3rd” “Orc.4th=Oct.4th”,…}, “Temperature”—>{“Oct.1st=23”, “Oct.2nd=14”, “Oct.4th=null”,…} }
		output: 2 —>{ “Date” —>{“Oct.1st=Oct.1st”, “Oct.2nd=Oct.2nd”, ”Oct.3rd=Oct.3rd” “Orc.4th=Oct.4th”,…}, “Temperature”—>{“Oct.1st=23”, “Oct.2nd=14”, “Oct.3rd=-1”“Oct.4th=0.0”,…} }
		(Oct.3rd column, not exists 	Oct.4th value is null)
    - this.data.put(target, dataItems)
6. Information about functions:
    - writeToMap
        * write List<JSONObject> records to local HashMap
        * Input: List<JSONObject>, Request request
        * Output: void, local Map<String, Map<String, Map<String, String>> data modified
    - fillMissingDateKeys
        * fill up the missing key and values of “Date”
        * Input: Map<String, Map<String, String>> dateItems, Request request
        * Output: void, dataItems modified
    - checkNullKeysAndValue
        * fill up the missing values in case of column not found and value is missing
        * Input: Map<String, Map<String, String>> dateItems, Request request
        * Output: void, dataItems modified

chrisyly commented on ucdavis_service/src/main/java/com/ucdavis/application/OpenWeatherService.java:
```diff
@@ -0,0 +1,67 @@
+ * @parm appKey: registered OpenWeatherMap APP KEY for services access
+ * @parm requestQuert: String query for OpenWeatherMap data
+ *  */
+public class OpenWeatherService {
```
This seems very similar to the cimis, do you think we can merge them later?

chrisyly commented on ucdavis_service/src/main/java/com/ucdavis/application/UCDavisService.java:
```diff
@@ -135,13 +142,87 @@ public void writeToMap(List<JSONObject> records, String target) {
 	}
 	
+	private void checkNullKeysAndValue(Map<String, Map<String, String>> dataItems, boolean isDaily) {
+		// daily keys
```
Daily check can be checked from json file, the Boolean variable should not be necessary

chrisyly reviewed (commented):
You pushed a lot of code, I do not have an environment to test now, can you test it and post what test you have done and the result?

DhYS committed:
Inserted the statement in JsonParser:
If inputString(considered as JSONArray) is null, return null Map<String, String>
